### DISCUSSION TITLE
How to configure relay host with only supported config files?

### DISCUSSION THREAD

**Discussion by pierinh00d** [docker-mailserver/docker-mailserver#3605] [Q&A]
Hi guys. I'm a new user of docker mailserver. 
I'm unable to work with relayhost setup. According to this tutorial https://docker-mailserver.github.io/docker-mailserver/edge/config/advanced/mail-forwarding/relay-hosts/#introduction RELAY* variables in env file are empty

My postfix-sasl-password.cf file:

`@domain1.it <username_hidden>:<password_hidden>`

and my postfix-relaymap.cf:

`@domain2.it`
`@domain1.it [email-smtp.eu-west-1.amazonaws.com]:587`
`@domain3.it`

But all email sent with @domain1.it aren't sent with relayhost
What am I doing wrong?
Thank you so much!

**Comment by polarathene** [comment]
> RELAY* variables in env file are empty

The docs have this big mention:

![image](https://github.com/docker-mailserver/docker-mailserver/assets/5098581/22a098d9-6375-4b9e-8ae0-1f04a21341ee)

In our scripts, you can see this requirement is currently required:

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L172-L195

You will need to create the real final configs used internally for Postfix, or enable `RELAY_HOST` for the DMS support to take your provided configs and create the internal ones that Postfix will use, otherwise this is skipped.

From what I can see:
- `RELAY_HOST` ENV will be used for any sender domain you haven't explicitly added into `postfix-relaymap.cf` (_to exclude or assign a relay host_).
- If you also set `RELAY_PASSWORD`, this may introduce a configuration issue if `RELAY_HOST` + `RELAY_PORT` are not configuring a valid relayhost. My advice is to consider `RELAY_PASSWORD` legacy and avoid setting it. Use your `postfix-sasl-password.cf` file only.
- Postfix does not appear to be configured with `relayhost` unless using the ENV `DEFAULT_RELAY_HOST`. However when you set `RELAY_HOST`, several related settings intended for relaying outbound mail are configured on the `smtp` transport (_uses the default `/etc/postfix/main.cf` config_) instead of the `relay` transport (_inherits `smtp` config and can override via `/etc/postfix/master.cf`_). Even without `relayhost` configured, these other changes may not be compatible for sending mail outbound without a relayhost destination.

Thus you can set `RELAY_HOST` to `email-smtp.eu-west-1.amazonaws.com` and `RELAY_PORT` to `587` and any domains you don't opt-out of via `postfix-relaymap.cf` would be routed through that relayhost. Similar to configuring `DEFAULT_RELAY_HOST=[email-smtp.eu-west-1.amazonaws.com]:587` (_except this will configure `relayhost` in Postfix `main.cf`, and lacks an opt-out AFAIK_), but presently at least `RELAY_HOST` must always be configured regardless.

**TL;DR:** You could try an empty `RELAY_HOST=" "`, that may work. So long as you don't trigger the two conditions that would try to use it as `[ ]:25`.

---

## Technical walk-through of current support implementation in DMS

When this feature is enabled it will set the following:

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L165-L170

- https://www.postfix.org/postconf.5.html#smtp_sasl_auth_enable
- https://www.postfix.org/postconf.5.html#smtp_tls_security_level

This is applying to all outbound mail from the looks of it, so DMS becomes a relayhost only and doesn't look like it supports sending mail outbound without going through a relay.

These two settings are related and at a glance imply a default configured `relayhost` would be used always (_unless the `sender_dependent_relayhost_maps` overrides that_):
- https://www.postfix.org/postconf.5.html#sender_dependent_relayhost_maps
- https://www.postfix.org/postconf.5.html#relayhost

I assume if `relayhost` is empty, it would allow you to specify only sender domains that should use a relayhost, however at least `smtp_tls_security_level` above might be an issue as it makes encryption mandatory which may cause problems, I've not verified.

---

Here we can see your `postfix-relaymap.cf` config being handled, it will copy the `@domain1.it` line to the real Postfix config `relayhost_map`:

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L109-L128

In that same method there is this TODO comment, that notes an improvement would be to use a default relay host, with support for opt-out of other sender domains, or the ability to adjust the transport (not implemented):

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L136-L139

---

For the SASL password config, you can see that it is requiring either`RELAY_HOST` or `RELAY_PASSWORD` is required:

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L60-L68

There is also some parts of this script file that reference a relayhost by calling this method:

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L47-L58

That will use `RELAY_HOST` + `RELAY_PORT` (_or default to port 25_), and this seems used as a fallback:

Legacy SASL password line added if both the `RELAY_*` ENV are configured (_not relevant when you configure `postfix-sasl-password` already_):

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L86-L89

The other usage looks at your `postfix-accounts.cf` (`setup email add ...`) and `postfix-virtual.cf` (`setup alias add ...`) for sender domains to automatically configure if they weren't already handled via `postfix-relaymap.cf`:

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L149-L160

Both are legacy and should not be relied on. As noted by comments, the current support is not done well:

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L99-L101

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L103-L108

---

Finally, that doc page I don't think has been updated really since it the migration from Github Wiki to the nicer docs site. There is another related [ENV called `DEFAULT_RELAY_HOST`](https://docker-mailserver.github.io/docker-mailserver/v12.1/config/environment/#default-relay-host):

https://github.com/docker-mailserver/docker-mailserver/blob/5efd2497864ab9bcf185b7901503786054b4a791/target/scripts/helpers/relay.sh#L175-L178

**Reply by polarathene** [reply]
I've opened a tracking issue for when this feature actually gets reworked: https://github.com/docker-mailserver/docker-mailserver/issues/3607

**Comment by polarathene** [comment]
> My `postfix-sasl-password.cf` file:
> 
> `@domain1.it <username_hidden>:<password_hidden>`
> 
> and my `postfix-relaymap.cf`:
> 
> ```
> @domain1.it [email-smtp.eu-west-1.amazonaws.com]:587
> ```
> 
> But all email sent with `@domain1.it` aren't sent through that relayhost, what am I doing wrong?

This will be supported (_no relay ENV settings required, just the relay config files_) with the upcoming v14 release: https://github.com/docker-mailserver/docker-mailserver/pull/3845

However, DMS still lacks support for the regular SMTP sending without credentials when a relay host has been configured with credentials.

**NOTE:** Quoted content has been altered slightly, notably `postfix-relaymap.cf` config to omit the other domains since no opt-out for them is required.
